apps/workers/openaiWorker.ts

```diff
@@ -166,6 +180,47 @@ async function inferTagsFromImage(
   );
 }
 
+async function inferTagsFromPDF(
+  jobId: string,
+  bookmark: NonNullable<Awaited<ReturnType<typeof fetchBookmark>>>,
+  inferenceClient: InferenceClient,
+) {
+  const { asset } = await readAsset({
+    userId: bookmark.userId,
+    assetId: bookmark.asset.assetId,
+  });
+  if (!asset) {
+    throw new Error(
+      `[inference][${jobId}] AssetId ${bookmark.asset.assetId} for bookmark ${bookmark.id} not found`,
+    );
+  }
+  const pdfParse = await readPDFText(asset);
+  if (!pdfParse?.text) {
+    throw new Error(
+      `[inference][${jobId}] PDF text is empty. Please make sure that the PDF includes text and not just images.`,
+    );
+  }
+
+  await db
+    .update(bookmarkAssets)
+    .set({
+      content: pdfParse.text,
+      info: pdfParse.info ? JSON.stringify(pdfParse.info) : null,
+      metadata: pdfParse.metadata ? JSON.stringify(pdfParse.metadata) : null,
+    })
+    .where(eq(bookmarkAssets.id, bookmark.id));
+
+  let prompt = PDF_PROMPT_BASE;
+  if (pdfParse.info) {
+    prompt += "Info: " + JSON.stringify(pdfParse.info) + "\n";
+  }
+  if (pdfParse.metadata) {
+    prompt += "Metadata: " + JSON.stringify(pdfParse.metadata) + "\n";
+  }
+  prompt += "Content: \n" + pdfParse.text;
```

Comment:
Models have a context window that we need to respect. You'll need to truncate the content to fit in that window especially if someone uploads a large pdf.

There's a function called `truncateContent` that's used in the text prompt builder, maybe you can reuse it?

Reply:
I had in mind the 4MB limit for the upload. But you are correct I'll add the truncate content.